federationtypes.go

```diff
@@ -256,6 +281,60 @@ func (r *RespMissingEvents) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+// MarshalJSON implements json.Marshaller
+func (r RespPeek) MarshalJSON() ([]byte, error) {
+	if len(r.StateEvents) == 0 {
+		r.StateEvents = []Event{}
+	}
+	if len(r.AuthEvents) == 0 {
+		r.AuthEvents = []Event{}
+	}
+	return json.Marshal(respPeekFields(r))
+}
+
+// UnmarshalJSON implements json.Unmarshaller
+func (r *RespPeek) UnmarshalJSON(data []byte) error {
+	r.AuthEvents = []Event{}
+	r.StateEvents = []Event{}
+	// XXX: it feels broken that we have RespPeek, respPeekFields and intermediate
```

Comment:
The key difference in the `intermediate` stage is that the state events and auth events are not unmarshalled on the first pass (hence `json.RawMessage`), which allows you to get `RoomVersion` so that you know what event format to unmarshal them as.